libindy/src/commands/metrics.rs

```diff
@@ -12,18 +29,585 @@ const OPENED_WALLET_IDS_COUNT: &str = "opened_wallet_ids_count";
+    // NonSecretsCommand
+    NonSecretsCommandAddRecord,
+    NonSecretsCommandUpdateRecordValue,
+    NonSecretsCommandUpdateRecordTags,
+    NonSecretsCommandAddRecordTags,
+    NonSecretsCommandDeleteRecordTags,
+    NonSecretsCommandDeleteRecord,
+    NonSecretsCommandGetRecord,
+    NonSecretsCommandOpenSearch,
+    NonSecretsCommandFetchSearchNextRecords,
+    NonSecretsCommandCloseSearch,
+    // PaymentsCommand
+    PaymentsCommandRegisterMethod,
+    PaymentsCommandCreateAddress,
+    PaymentsCommandCreateAddressAck,
+    PaymentsCommandListAddresses,
+    PaymentsCommandAddRequestFees,
+    PaymentsCommandAddRequestFeesAck,
+    PaymentsCommandParseResponseWithFees,
+    PaymentsCommandParseResponseWithFeesAck,
+    PaymentsCommandBuildGetPaymentSourcesRequest,
+    PaymentsCommandBuildGetPaymentSourcesRequestAck,
+    PaymentsCommandParseGetPaymentSourcesResponse,
+    PaymentsCommandParseGetPaymentSourcesResponseAck,
+    PaymentsCommandBuildPaymentReq,
+    PaymentsCommandBuildPaymentReqAck,
+    PaymentsCommandParsePaymentResponse,
+    PaymentsCommandParsePaymentResponseAck,
+    PaymentsCommandAppendTxnAuthorAgreementAcceptanceToExtra,
+    PaymentsCommandBuildMintReq,
+    PaymentsCommandBuildMintReqAck,
+    PaymentsCommandBuildSetTxnFeesReq,
+    PaymentsCommandBuildSetTxnFeesReqAck,
+    PaymentsCommandBuildGetTxnFeesReq,
+    PaymentsCommandBuildGetTxnFeesReqAck,
+    PaymentsCommandParseGetTxnFeesResponse,
+    PaymentsCommandParseGetTxnFeesResponseAck,
+    PaymentsCommandBuildVerifyPaymentReq,
+    PaymentsCommandBuildVerifyPaymentReqAck,
+    PaymentsCommandParseVerifyPaymentResponse,
+    PaymentsCommandParseVerifyPaymentResponseAck,
+    PaymentsCommandGetRequestInfo,
+    PaymentsCommandSignWithAddressReq,
+    PaymentsCommandSignWithAddressAck,
+    PaymentsCommandVerifyWithAddressReq,
+    PaymentsCommandVerifyWithAddressAck,
+    // CacheCommand
+    CacheCommandGetSchema,
+    CacheCommandGetSchemaContinue,
+    CacheCommandGetCredDef,
+    CacheCommandGetCredDefContinue,
+    CacheCommandPurgeSchemaCache,
+    CacheCommandPurgeCredDefCache,
+    // MetricsCommand
+    MetricsCommandCollectMetrics,
+    // Exit
+    Exit,
+}
+
+
 pub enum MetricsCommand {
     CollectMetrics(Box<dyn Fn(IndyResult<String>) + Send>)
 }
 
+const COMMANDS_COUNT: usize = MetricsStorage::commands_count();
+
+pub struct MetricsStorage {
+    queue_count: [u64; COMMANDS_COUNT],
+    queue_duration: [u64; COMMANDS_COUNT],
+
+    execute_count: [u64; COMMANDS_COUNT],
+    execute_duration: [u64; COMMANDS_COUNT],
+}
+
+impl MetricsStorage {
+    pub fn new() -> MetricsStorage {
+        MetricsStorage {
+            queue_count: [u64::MIN; COMMANDS_COUNT],
+            queue_duration: [u64::MIN; COMMANDS_COUNT],
+
+            execute_count: [u64::MIN; COMMANDS_COUNT],
+            execute_duration: [u64::MIN; COMMANDS_COUNT],
+        }
+    }
+
+    pub fn cmd_left_queue(&mut self, command_index: usize, duration: u64) {
```

Comment:
Why not just pass `Command` here, or at least `CommandIndex`? It would be much less error prone.

Btw, if you try to avoid redundant conversions from `Command` to `CommandIndex` to `usize` to get the best performance possible then:
* I'm 99% sure conversion from `CommandIndex` to `usize` is zero-cost
* I'm 90% sure conversion from `Command` to `CommandIndex` is close to zero-cost

Reply:
No, I am struggling with mutable links issue. But I hope to deal with this problem today. Therefore, there is still TODO.